code/game/machinery/autolathe.dm

```diff
@@ -123,18 +123,21 @@
 	return ..()
 
 
-/obj/machinery/autolathe/proc/AfterMaterialInsert(item_inserted, id_inserted, amount_inserted)
+/obj/machinery/autolathe/proc/AfterMaterialInsert(obj/item/item_inserted, id_inserted, amount_inserted)
 	if(istype(item_inserted, /obj/item/stack/ore/bluespace_crystal))
 		use_power(MINERAL_MATERIAL_AMOUNT / 10)
-	else if(custom_materials && custom_materials.len && custom_materials[SSmaterials.GetMaterialRef(/datum/material/glass)])
-		flick("autolathe_r",src)//plays glass insertion animation by default otherwise
+	else if(item_inserted.has_material_type(/datum/material/glass))
+		flick("autolathe_r", src)//plays glass insertion animation by default otherwise
 	else
-		flick("autolathe_o",src)//plays metal insertion animation
-
+		flick("autolathe_o", src)//plays metal insertion animation
 
 		use_power(min(1000, amount_inserted / 100))
 	updateUsrDialog()
 
+/// Checks whether a material is rigid
```

Comment:
Why does this have to be handled through a callback? Seems like it should just be abled to handle that through the fact that we requested rigid mats to be accepted by it.

Reply:
I'd love if it worked that way, but that's really not how material containers work. The material containers use the list of materials they contain to check whether a material fits into the container. If it isn't on the list, it doesn't fit. I could change that, but that would require more callbacks and more complicated code for just inserting objects.

If you want me to I'll do it.

Reply:
On the other hand, reworking material containers like that would make instantiating new bespoke materials not O(n) in relation to the number of autolathes in existance.

Reply:
Done

Reply:
I think you're missing my point. We literaly tell it to use materials from SSmaterials.materialids_by_category[MAT_CATEGORY_RIGID]. This means we want it to accept all materials with MAT_CATEGORY_RIGID, but then we also have a callback which checks if the material is rigid. This means it's doing double work. Why not just make it so you pass MAT_CATEGORY_RIGID in the component creation and then the component checks if the material youre trying to put in has MAT_CATEGORY_RIGID. (Preferably, this would support lists so you could also say you only want materials that are both MAT_CATEGORY_RIGID and MAT_CATEGORY_ORE

Reply:
I would like to be able to do this, the issue is that I have no idea what spaghetti expandable sets of materials future coders will want to have their machines to be able to contain. Ergo a callback for the cases where someone wants everything in three material categories that isn't in another two categories except actually contain this type but not these five for reasons.

Reply:
that does not change that in most cases, you shouldnt have to make a callback. You can add a callback as an optional thing but it should not be required for every implementation that uses mat categories.

Reply:
Done. The callback is now optional.